src/middlewared/middlewared/plugins/alert.py

```diff
@@ -0,0 +1,444 @@
+                for gone_alert in (set(self.last_key_value_alerts[alert_source_name].keys()) -
+                                   set(alerts[alert_source_name].keys())):
+                    gone_alerts.append(self.last_key_value_alerts[alert_source_name][gone_alert])
+                for new_alert in (set(alerts[alert_source_name].keys()) -
+                                  set(self.last_key_value_alerts[alert_source_name].keys())):
+                    new_alerts.append(alerts[alert_source_name][new_alert])
+
+            self.last_key_value = key
+            self.last_key_value_alerts = copy.deepcopy(alerts)
+
+        return gone_alerts, new_alerts
+
+
+class AlertService(Service):
+    def __init__(self, middleware):
+        super().__init__(middleware)
+
+        self.node = "A"
+
+        self.alerts = defaultdict(dict)
+
+        self.alert_source_last_run = defaultdict(lambda: datetime.min)
+
+        self.policies = {
+            "IMMEDIATELY": AlertPolicy(),
+            "HOURLY": AlertPolicy(lambda d: (d.date(), d.hour)),
+            "DAILY": AlertPolicy(lambda d: (d.date())),
+            "NEVER": AlertPolicy(lambda d: None),
+        }
+
+    @private
+    async def initialize(self):
+        if not await self.middleware.call("system.is_freenas"):
+            if await self.middleware.call("notifier.failover_node") == "B":
+                self.node = "B"
+
+        for alert in await self.middleware.call("datastore.query", "system.alert", [("node", "=", self.node)]):
+            del alert["id"]
+            del alert["node"]
+            alert["level"] = AlertLevel(alert["level"])
+
+            alert = Alert(**alert)
+
+            self.alerts[alert.source][alert.key] = alert
+
+        for policy in self.policies.values():
+            policy.receive_alerts(datetime.utcnow(), self.alerts)
+
+        for module in load_modules(os.path.join(os.path.dirname(os.path.realpath(__file__)), os.path.pardir,
+                                                "alert", "source")):
+            for cls in load_classes(module, AlertSource, (FilePresenceAlertSource, ThreadedAlertSource)):
+                alert_sources.append(cls(self.middleware))
+
+        for module in load_modules(os.path.join(os.path.dirname(os.path.realpath(__file__)), os.path.pardir,
+                                                "alert", "service")):
+            for cls in load_classes(module, _AlertService, (ThreadedAlertService,)):
+                alert_services_factories[cls.name()] = cls
+
+    async def terminate(self):
+        await self.flush_alerts()
+
+    async def list_policies(self):
+        return POLICIES
+
+    async def list_sources(self):
+        return [
+            {
+                "name": source.name,
+                "title": source.title,
+            }
+            for source in sorted(alert_sources, key=lambda source: source.title.lower())
+        ]
+
+    def list(self):
```

Comment:
We need @accepts() to make these methods public and show up in the docs.

Some docstring on them would be greatly appreciate as well to help UI devs.